### ISSUE TITLE
[Epic] Casting Column Types for Postgres Sources

### ISSUE THREAD

**Issue by nmeagan11** [MaterializeInc/materialize#15716]
### Feature request

In https://github.com/MaterializeInc/materialize/issues/14016, we removed `CREATE VIEWS`. This left us with the limitation that if a user has a table in the publication that contains an unsupported column type, the source will crash. We need to add a more fine-grained control as a workaround.

Proposal for the UX (stolen from @sjwiesman because I like it):

```
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
      PUBLICATION 'mz_source',
      CAST foo.my_enum AS string)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

We also need to consider the case where the user fails to specify all columns that contain an unsupported type. In this case, I like @benesch's previous suggestion that we provide an error message that tells the user how they can fix their `CREATE SOURCE` statement, e.g.:

```
ERROR: Column "foo" in table "bar" contains an unsupported type. Please use CAST in the CREATE SOURCE statement.
```

### QA Tasks

TBD

**Comment by sjwiesman** [comment]
I'm certainly onboard with the current proposal.

**Comment by ggnall** [comment]
Error messages should definitely be instructive! See style guide here: https://github.com/chaas/materialize/blob/83444c57a5f1f414e377b5decb432058e1233b92/doc/developer/style.md#sql-error-style

**Comment by nmeagan11** [comment]
> Error messages should definitely be instructive! See style guide here

Nice! I didn't know that guide existed. I made a small update, but someone that's an expert should probably validate and/or improve.

**Comment by benesch** [comment]
How does the syntax look with multiple casts? Also I think we probably should push the `CAST` option into the parenthesis, so it's affiliated with all the other Postgres-specific options:

```
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    CAST foo.my_enum AS string,
    CAST foo.my_other_enum AS string
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

**Comment by nmeagan11** [comment]
> we probably should push the CAST option into the parenthesis

Makes sense! Edited above.

> multiple casts

I don't think you can get around just listing them out? Unless you're getting at something else?

**Comment by ggnall** [comment]
Recommendation for error message. cc @chaas 
```sql
Primary: could not parse column "foo"
Detail: Column "foo" in table "bar" contains an unsupported type.
Hint: Try using CAST in the CREATE SOURCE statement.
```

**Comment by petrosagg** [comment]
Is there any case where the `CAST` syntax makes sense with the target type being something other than string? If all usages are "just leave column this as string" maybe we can improve the syntax by not requiring a target type?

Strawman:

```sql
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    WITH RAW COLUMNS foo.my_enum, foo.my_other_enum
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');

**Comment by sploiselle** [comment]
Assuming we do just want to leave the input as strings, how would you all feel about making a blanket version of this?

```sql
REATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    ALLOW ALL UNSUPPORTED TYPES
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

and then let you also list them individually as Petros suggests:

```sql
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    ALLOW UNSUPPORTED TYPES (foo.my_enum, foo.my_other_enum)
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

**Comment by petrosagg** [comment]
I think the blanket one has forward compatibility issues if we add more types because a user re-creating the source with the same statement and upstream schema will get different a different materialized schema

**Comment by sjwiesman** [comment]
> Is there any case where the CAST syntax makes sense with the target type being something other than string? If all usages are "just leave column this as string" maybe we can improve the syntax by not requiring a target type?

I suspect you are right about casting to string for this source, but would be in favor of retainng `CAST .. AS ..` for the syntax so the syntax is consistent with theoretical future sources. We plan to support other direct database sources, and those type systems may not map 1-1 with postgres. In those instances, users will need the ability to specify types other than STRING.

**Comment by sploiselle** [comment]
@sjwiesman Can you think of any concrete examples where a representation other than string is reasonable?

**Comment by petrosagg** [comment]
The target type doesn't have to be a string necessarily. The way I think about it that we provide the users with a switch for each column. That switch is "ingest this column as a different type than it is". What type we choose to use can be something richer than just picking `text` unconditionally. Our docs could say "When you enable `ALLOW UNSUPPORTED` for an enum column, it will be ingested as `int64`. For a `foo` column it will be ingested as `int32`, and for everything else as `text`.

So the question more generally is if we envision a case where an upstream source has some type that we don't support and at the same time can be reasonably cast into two different supported mz types and we want the user to be able to choose between them.

**Comment by sploiselle** [comment]
Ah, enums were an example of what I was asking after. So we probably want PG enums to be integers, but we also only get once chance to define these alternate types once we release support for a source. That seems hard to get exactly right for all time, and prevents us from moving from the original decision to some other supported type.

What if we:
- Offer `ALLOW UNSUPPORTED` for this default behavior (e.g. everything to text except enums to ints) as a v0 of the feature
- Can add `CAST unsupported_type AS supported_type` if anyone wants any other behavior, and `ALLOW UNSUPPORTED` essentially becomes just syntactic sugar over a series of `CAST unsupported_type AS default` options.

**Comment by benesch** [comment]
I like the idea of `WITH RAW COLUMNS`, though I'd propose naming the option `RAW COLUMNS` or `TEXT COLUMNS` instead:

```sql
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    TEXT COLUMNS (foo.my_enum, foo.my_other_enum)
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

> * Can add `CAST unsupported_type AS supported_type` if anyone wants any other behavior, and `ALLOW UNSUPPORTED` essentially becomes just syntactic sugar over a series of `CAST unsupported_type AS default` options.

Thinking about this more... I think we can probably punt on this! We're likely to support attaching arbitrary `SELECT` clauses to sources so that you can specify an MFP that should be pushed down into ingestion before the data is persisted. So if we build that more general feature, you'll be able to put your cast expressions there.

**Comment by petrosagg** [comment]
`TEXT COLUMNS` hits the sweet spot for me of concise syntax and self-documentation where you can tell the target type just by looking at the query. I cast my vote!

**Comment by sploiselle** [comment]
Ah, I hadn't heard of attaching MFPs to sources--that's fantastic. Agreed on `TEXT COLUMNS`.

**Comment by morsapaes** [comment]
`CAST UNSUPPORTED` could be more descriptive than `TEXT COLUMNS`. In this case, I think it's more intuitive to focus on what the clause does rather than the outcome. Then we can document that the default cast type is `text`.

```sql
CREATE SOURCE mz_source
FROM POSTGRES CONNECTION pg_connection (
    PUBLICATION 'mz_source',
    CAST UNSUPPORTED (foo.my_enum, foo.my_other_enum)
)
FOR ALL TABLES
WITH (SIZE = '3xsmall');
```

**Comment by ggnall** [comment]
We met and decided to go forward with `TEXT COLUMNS`. Updating error message to match:
```
Primary: could not parse column "foo.bar"
Detail: Column "foo.bar" contains an unsupported type.
Hint: Try using TEXT COLUMNS in the CREATE SOURCE statement.
```

**Comment by benesch** [comment]
Minor nit: I don't think "parse" is quite the right verb; it's not that the data in the column failed to parse but rather that we don't have a parser to use. Maybe something like this?

```
ERROR: column "foo.bar" contains an unsupported type
Detail: The type has OID 12345 in the upstream database.
Hint: Try using TEXT COLUMNS in the CREATE SOURCE statement.
```

Or... maybe even better is to list all the unsupported types at once! That way you don't have to discover them one by one.

```
ERROR:  source contains a table with an unsupported type
Detail: The following columns have unsupported types:
    foo.bar (OID: 1234)
    foo.bar2 (OID: 1234)
    baz.bas (OID: 4321)
Hint: Try using the TEXT COLUMNS option in the CREATE SOURCE statement.
```

**Comment by ggnall** [comment]
Yes! The list version is $
> Or... maybe even better is to list all the unsupported types at once! That way you don't have to discover them one by one.

**Comment by chaas** [comment]
> Detail: The following columns have unsupported types:
    foo.bar (OID: 1234)
    foo.bar2 (OID: 1234)
    baz.bas (OID: 4321)

One note: from the Postgres error style guide

> In long messages, newline characters (\n) can be used to indicate suggested paragraph breaks. Don't end a message with a newline. Don't use tabs or other formatting characters. 

So we should either use spaces for the indents or avoid them altogether in the formatting of the Detail and Hint messages.

**Comment by sploiselle** [comment]
@nmeagan11 @philip-stoev added tests to the merged PR #16010 , as well as approved
